src/dataflow/src/source/timestamp.rs

```diff
@@ -0,0 +1,259 @@
+        assert!(
+            offset >= *last_offset,
+            "offset should not go backwards, but {} < {}",
+            offset,
+            last_offset
+        );
+        assert!(
+            timestamp > *last_ts,
+            "timestamp should move forwards, but {} <= {}",
+            timestamp,
+            last_ts
+        );
+        entry.push((timestamp, offset));
+    }
+
+    fn get_binding(
+        &self,
+        partition: &PartitionId,
+        offset: MzOffset,
+    ) -> Option<(Timestamp, MzOffset)> {
+        if !self.partitions.contains_key(partition) {
+            return None;
+        }
+
+        let entries = self.partitions.get(partition).expect("known to exist");
+
+        for (ts, max_offset) in entries {
+            if offset <= *max_offset {
+                return Some((*ts, *max_offset));
+            }
+        }
+
+        return None;
+    }
+
+    fn read_upper(&self, target: &mut Antichain<Timestamp>) {
+        target.clear();
+
+        for (_, entries) in self.partitions.iter() {
+            if let Some((timestamp, _)) = entries.last() {
+                target.insert(*timestamp);
+            }
+        }
+    }
+
+    fn partitions(&self) -> Vec<PartitionId> {
+        self.partitions
+            .iter()
+            .map(|(pid, _)| pid)
+            .cloned()
+            .collect()
+    }
+}
+
+/// A wrapper that allows multiple source instances to share a `TimestampBindingBox`
+/// and hold back its compaction.
+pub struct TimestampBindingRc {
+    wrapper: Rc<RefCell<TimestampBindingBox>>,
+    compaction_frontier: Antichain<Timestamp>,
+}
+
+impl TimestampBindingRc {
+    /// Create a new instance of `TimestampBindingRc`.
+    pub fn new() -> Self {
+        let wrapper = Rc::new(RefCell::new(TimestampBindingBox::new()));
+
+        let ret = Self {
+            wrapper: wrapper.clone(),
+            compaction_frontier: wrapper.borrow().compaction_frontier.frontier().to_owned(),
+        };
+
+        ret
+    }
+
+    /// Set the compaction frontier to `new_frontier`.
+    ///
+    /// Note that `new_frontier` must be in advance of the current compaction
+    /// frontier.
+    pub fn set_compaction_frontier(&mut self, new_frontier: AntichainRef<Timestamp>) {
+        self.wrapper
+            .borrow_mut()
+            .adjust_compaction_frontier(self.compaction_frontier.borrow(), new_frontier);
+        self.compaction_frontier = new_frontier.to_owned();
+    }
+
+    /// Compact all timestamp bindings at timestamps less than the current
+    /// compaction frontier.
+    ///
+    /// The source can be correctly replayed from any `as_of` in advance of
+    /// the compaction frontier after this operation.
+    pub fn compact(&self) {
+        self.wrapper.borrow_mut().compact();
+    }
+
+    /// Add a new mapping from `(partition, offset) -> timestamp`.
+    ///
+    /// Note that the `timestamp` greater than the largest previously bound
```

Comment:
missing a "must be" I think?

Reply:
Minor stylistic nit: you can probably ditch 99% of the occurrences of "Note that".